src/test/java/io/numaproj/numaflow/mapper/ServerTest.java

```diff
@@ -57,31 +59,86 @@ public void tearDown() throws Exception {
         ByteString expectedValue = ByteString.copyFromUtf8("invalue" + PROCESSED_VALUE_SUFFIX);
 
+        MapOutputStreamObserver responseObserver = new MapOutputStreamObserver(4);
+
+        var stub = MapGrpc.newStub(inProcessChannel);
+        var requestStreamObserver = stub
+                .mapFn(responseObserver);
+
+        requestStreamObserver.onNext(handshakeRequest);
+        requestStreamObserver.onNext(inDatum);
+        requestStreamObserver.onNext(inDatum);
+        requestStreamObserver.onNext(inDatum);
+
+        try {
+            responseObserver.done.get();
+        } catch (InterruptedException | ExecutionException e) {
+            fail("Error while waiting for response" + e.getMessage());
+        }
+
+        List<MapOuterClass.MapResponse> responses = responseObserver.getMapResponses();
+        assertEquals(4, responses.size());
 
-        var stub = MapGrpc.newBlockingStub(inProcessChannel);
-        var mapResponse = stub
-                .mapFn(inDatum);
+        // first response is the handshake response
+        assertEquals(handshakeRequest.getHandshake(), responses.get(0).getHandshake());
 
-        assertEquals(1, mapResponse.getResultsCount());
-        assertEquals(
-                expectedKeys,
-                mapResponse.getResults(0).getKeysList().toArray(new String[0]));
-        assertEquals(expectedValue, mapResponse.getResults(0).getValue());
-        assertEquals(
-                expectedTags,
-                mapResponse.getResults(0).getTagsList().toArray(new String[0]));
+        responses = responses.subList(1, responses.size());
+        for (MapOuterClass.MapResponse response : responses) {
+            assertEquals(expectedValue, response.getResults(0).getValue());
+            assertEquals(Arrays.asList(expectedKeys), response.getResults(0).getKeysList());
+            assertEquals(Arrays.asList(expectedTags), response.getResults(0).getTagsList());
+            assertEquals(1, response.getResultsCount());
+        }
+
+        requestStreamObserver.onCompleted();
+    }
+
+    @Test
+    public void testMapperWithoutHandshake() {
+        ByteString inValue = ByteString.copyFromUtf8("invalue");
+        MapOuterClass.MapRequest inDatum = MapOuterClass.MapRequest
+                .newBuilder()
+                .setRequest(MapOuterClass.MapRequest.Request
+                        .newBuilder()
+                        .setValue(inValue)
+                        .addAllKeys(List.of("test-map-key"))
+                        .build()).build();
+
+        MapOutputStreamObserver responseObserver = new MapOutputStreamObserver(1);
+
+        var stub = MapGrpc.newStub(inProcessChannel);
+        var requestStreamObserver = stub
+                .mapFn(responseObserver);
+
+        requestStreamObserver.onNext(inDatum);
+
+        try {
+            responseObserver.done.get();
+        } catch (InterruptedException | ExecutionException e) {
```

Comment:
nit: if there was no exception thrown, this unit test could pass. consider returning within the catch block and failing if reaching the end of the test.